src/app/utilities/transmart-utilities/transmart-constraint-reader.ts

```diff
@@ -0,0 +1,321 @@
+import {AbstractTransmartConstraintVisitor} from './abstract-transmart-constraint-visitor';
+import {Constraint} from '../../models/constraint-models/constraint';
+import {
+  ExtendedAndConstraint,
+  ExtendedConceptConstraint, ExtendedConstraint,
+  TransmartCombinationConstraint,
+  TransmartFieldConstraint,
+  TransmartNegationConstraint,
+  TransmartNullConstraint,
+  TransmartPatientSetConstraint,
+  TransmartRelationConstraint,
+  TransmartStudyNameConstraint,
+  TransmartSubSelectionConstraint,
+  TransmartTemporalConstraint,
+  TransmartTimeConstraint,
+  TransmartTrueConstraint,
+  TransmartValueConstraint
+} from '../../models/transmart-models/transmart-constraint';
+import {Concept} from '../../models/constraint-models/concept';
+import {ConceptConstraint} from '../../models/constraint-models/concept-constraint';
+import {Study} from '../../models/constraint-models/study';
+import {StudyConstraint} from '../../models/constraint-models/study-constraint';
+import {CombinationConstraint} from '../../models/constraint-models/combination-constraint';
+import {CombinationState} from '../../models/constraint-models/combination-state';
+import {TimeConstraint} from '../../models/constraint-models/time-constraint';
+import {TrialVisitConstraint} from '../../models/constraint-models/trial-visit-constraint';
+import {ValueConstraint} from '../../models/constraint-models/value-constraint';
+import {PedigreeConstraint} from '../../models/constraint-models/pedigree-constraint';
+import {TrialVisit} from '../../models/constraint-models/trial-visit';
+import {SubjectSetConstraint} from '../../models/constraint-models/subject-set-constraint';
+import {TrueConstraint} from '../../models/constraint-models/true-constraint';
+import {DateOperatorState} from '../../models/constraint-models/date-operator-state';
+import {Operator} from '../../models/constraint-models/operator';
+
+/**
+ * Deserialisation class for reading constraint objects from TranSMART API and
+ * transform them into constraints that can be used in the cohort selection.
+ *
+ * To deserialise a constraint, use <code>new TransmartConstraintReader().visit(constraint)</code>.
+ */
+export class TransmartConstraintReader extends AbstractTransmartConstraintVisitor<Constraint> {
```

Comment:
as I understand, CohortMapper does both serialisation and deserialisation for cohorts, one would expect there is also a TransmartConstraintMapper that does serialisation and deserialisation for transmart constraints, right?

Reply:
There are two reasons for this: 
1. the transmart constraint mappers implement the visitor interfaces
2. size